apps/src/studio/studio.js

```diff
@@ -1894,43 +1884,56 @@ Studio.reset = function(first) {
   for (i = 0; i < Studio.spriteCount; i++) {
     Studio.displaySprite(i);
   }
+  this.resetGoalSprites();
+  sortDrawOrder();
 
-  var goalAsset = skin.goal;
-  if (level.goalOverride) {
-    if (level.goalOverride.goalAnimation) {
-      goalAsset = skin[level.goalOverride.goalAnimation];
-    } else if (level.goalOverride.goalImage) {
-      goalAsset = skin[level.goalOverride.goalImage];
-    }
-  }
+  // A little flag for script-based code to consume.
+  Studio.levelRestarted = true;
 
+  // Reset whether level has succeeded.
+  Studio.succeededTime = null;
+};
+
+/**
+ * Move all goal sprites to their original positions, and reset their completion
+ * state, both visual and logical.
+ */
+Studio.resetGoalSprites = function () {
   Studio.touchAllGoalsEventFired = false;
-  for (i = 0; i < Studio.spriteGoals_.length; i++) {
+  for (var i = 0; i < Studio.spriteGoals_.length; i++) {
     // Mark each finish as incomplete.
     Studio.spriteGoals_[i].finished = false;
     Studio.spriteGoals_[i].startFadeTime = null;
 
     // Move the finish icons into position.
-    var offsetX = skin.goalRenderOffsetX || 0;
-    var offsetY = skin.goalRenderOffsetY || 0;
+    var goalOverride = utils.valueOr(level.goalOverride, {});
+    var offsetX = utils.valueOr(goalOverride.goalRenderOffsetX,
+        utils.valueOr(skin.goalRenderOffsetX, 0));
+    var offsetY = utils.valueOr(goalOverride.goalRenderOffsetY,
+        utils.valueOr(skin.goalRenderOffsetY, 0));
     var spriteFinishIcon = document.getElementById('spriteFinish' + i);
     spriteFinishIcon.setAttribute('x', Studio.spriteGoals_[i].x + offsetX);
     spriteFinishIcon.setAttribute('y', Studio.spriteGoals_[i].y + offsetY);
     spriteFinishIcon.setAttributeNS('http://www.w3.org/1999/xlink',
-      'xlink:href', goalAsset);
+        'xlink:href', Studio.getGoalAssetFromSkin());
```

Comment:
`goalAsset` was originally calculated 20 lines ago and only used once here; now extracted into a helper method below.